Lib/subprocess.py

```diff
@@ -1480,7 +1480,26 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             if shell:
                 startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
                 startupinfo.wShowWindow = _winapi.SW_HIDE
-                comspec = os.environ.get("COMSPEC", "cmd.exe")
+                # gh-101283: without a fully-qualified path, before Windows
+                # checks the system directories, it first looks in the
+                # application directory, and also the current directory if
+                # NeedCurrentDirectoryForExePathW(ExeName) is true, so try
+                # to avoid executing unqualified "cmd.exe".
+                comspec = os.environ.get('ComSpec')
```

Comment:
nit: are environment variables case insensitive on Windows?  is that true within os.environ?  This is not mentioned in https://docs.python.org/3/library/os.html#os.environ.

Reply:
Yes, environment variable names are case insensitive and case preserving when accessed via the API functions `GetEnvironmentVariableW()` and `SetEnvironmentVariableW()`. 

The `os.environ` mapping implements case-insensitive access by storing the keys as uppercase strings. It sets environment variables via `os.putenv()`, which ultimately calls `SetEnvironmentVariableW()`. The latter preserves the original name when updating the value of an existing variable.

A small issue, which is mostly a matter of aesthetics, concerns using `os.environ` as the `env` parameter of `subprocess.Popen`. The child process gets an environment block with uppercase variable names. We could implement `os.environ` more cleverly to get a case insensitive, case preserving dict.